From 6df863141b0fcd945b940e82d9db78ab3ea7a4a5 Mon Sep 17 00:00:00 2001
From: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Date: Tue, 18 Jan 2022 17:06:07 +0100
Subject: [PATCH 149/841] feat(stm32mp2): save/restore DDR PHY training results
 for low power modes

After first DDR init, registers containing the training results are saved
(only as global variables here, but in a retention-enabled memory area at
term). Exiting from standby low power mode, DDR is initialized without
training, and those registers are then restored.
Remove STM32MP_DDR_SKIP_TRAINING flag use, both cases are needed in our way
to treat nominal and low power modes.
Fit MAX_NUM_RET_REGS with real needs.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 60d9eff6b0811d8f5973a3d99acea4a3ff283363 feat(stm32mp2): save/restore DDR PHY training results for low power modes]
Signed-off-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Change-Id: I7c2e451671110566c2e89e550f056880f64587f5
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/236252
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 .../include/ddrphy_phyinit_usercustom.h       |  4 ++--
 .../src/ddrphy_phyinit_i_loadpieimage.c       |  4 ----
 .../phyinit/src/ddrphy_phyinit_reginterface.c |  8 ++++++--
 .../phy/phyinit/src/ddrphy_phyinit_sequence.c | 20 +++++++++----------
 drivers/st/ddr/stm32mp2_ddr.c                 | 15 +++++++++++---
 plat/st/stm32mp2/platform.mk                  |  5 ++++-
 6 files changed, 33 insertions(+), 23 deletions(-)

diff --git a/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_usercustom.h b/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_usercustom.h
index 4f6829e23..0e4626b58 100644
--- a/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_usercustom.h
+++ b/drivers/st/ddr/phy/phyinit/include/ddrphy_phyinit_usercustom.h
@@ -103,7 +103,7 @@ typedef enum { return_offset, return_lastaddr } return_offset_lastaddr_t;
 /* A structure to store the sequence function runtime input variables. */
 typedef struct runtime_config {
 	bool	skip_train;	/* skip_train input parameter */
-	int	reten;		/*
+	bool	reten;		/*
 				 * Retention Enable input parameter, instructs phyinit to issue
 				 * register reads during initialization to retention registers.
 				 */
@@ -148,7 +148,7 @@ extern int ardptrinitval[NB_PS];
  * Fixed Function prototypes
  * -------------------------------------------------------------
  */
-int ddrphy_phyinit_sequence(bool skip_training);
+int ddrphy_phyinit_sequence(bool skip_training, bool reten);
 int ddrphy_phyinit_restore_sequence(void);
 void ddrphy_phyinit_c_initphyconfig(void);
 void ddrphy_phyinit_d_loadimem(void);
diff --git a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_i_loadpieimage.c b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_i_loadpieimage.c
index 862b2702e..6c2a6e98d 100644
--- a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_i_loadpieimage.c
+++ b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_i_loadpieimage.c
@@ -156,7 +156,6 @@ static void seq0bdisableflag_program(bool skip_training)
 }
 
 #if STM32MP_LPDDR4_TYPE
-#if !STM32MP_DDR_SKIP_TRAINING
 /*
  * Registers: ppttrainsetup and ppttrainsetup2
  * - Related to DFI PHY Master Interface (PMI). See Register description in PUB.
@@ -197,7 +196,6 @@ static void ppttrainsetup_program(void)
 		}
 	}
 }
-#endif /* !STM32MP_DDR_SKIP_TRAINING */
 
 /*
  * Registers AcsmPlayback*x*
@@ -392,11 +390,9 @@ void ddrphy_phyinit_i_loadpieimage(bool skip_training)
 	seq0bdisableflag_program(skip_training);
 
 #if STM32MP_LPDDR4_TYPE
-#if !STM32MP_DDR_SKIP_TRAINING
 	if (!skip_training) {
 		ppttrainsetup_program();
 	}
-#endif /* !STM32MP_DDR_SKIP_TRAINING */
 
 	acsmplayback_program();
 
diff --git a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_reginterface.c b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_reginterface.c
index d18da7f69..4b5204762 100644
--- a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_reginterface.c
+++ b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_reginterface.c
@@ -28,7 +28,11 @@
  * amount of registered that can be saved. The user may increase this variable
  * as desired if a larger number of registers need to be restored.
  */
-#define MAX_NUM_RET_REGS 5000
+#if STM32MP_DDR3_TYPE || STM32MP_DDR4_TYPE
+#define MAX_NUM_RET_REGS	416
+#elif STM32MP_LPDDR4_TYPE
+#define MAX_NUM_RET_REGS	281
+#endif /* STM32MP_LPDDR4_TYPE */
 
 /*
  * Array of Address/value pairs used to store register values for the purpose
@@ -70,7 +74,7 @@ int ddrphy_phyinit_trackreg(uint32_t adr)
 	}
 
 	if (!foundreg && tracken) { /* register not found, so add it. */
-		if (numregsaved == MAX_NUM_RET_REGS) {
+		if (numregsaved > MAX_NUM_RET_REGS) {
 			ERROR("[ddrphy_phyinit_reginterface:ddrphy_phyinit_trackreg]\n");
 			ERROR("Max Number of Restore Registers reached: %d.\n", numregsaved);
 			ERROR("Please recompile PhyInit with larger MAX_NUM_RET_REG value.\n");
diff --git a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_sequence.c b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_sequence.c
index 98f7c0b48..4a84a3ccd 100644
--- a/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_sequence.c
+++ b/drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_sequence.c
@@ -17,7 +17,7 @@
  *
  * \returns 0 on completion of the sequence, EXIT_FAILURE on error.
  */
-int ddrphy_phyinit_sequence(bool skip_training)
+int ddrphy_phyinit_sequence(bool skip_training, bool reten)
 {
 	/* Check user input pstate number consistency vs. SW capabilities */
 	if (userinputbasic.numpstates > NB_PS) {
@@ -53,13 +53,10 @@ int ddrphy_phyinit_sequence(bool skip_training)
 	/* Stop retention register tracking for training firmware related registers */
 	ddrphy_phyinit_reginterface(stoptrack, 0, 0);
 
-#if STM32MP_DDR_SKIP_TRAINING
-	if (skip_training) {
+	if (runtimeconfig.skip_train) {
 		/* Skip running training firmware entirely */
-		ddrphy_phyinit_progcsrskiptrain(skip_training);
-	}
-#else
-	{
+		ddrphy_phyinit_progcsrskiptrain(runtimeconfig.skip_train);
+	} else {
 		int pstate;
 
 		/* Run all 1D power states, then 2D P0, to reduce total Imem/Dmem loads. */
@@ -89,13 +86,12 @@ int ddrphy_phyinit_sequence(bool skip_training)
 			ddrphy_phyinit_h_readmsgblock();
 		}
 	}
-#endif /* STM32MP_DDR_SKIP_TRAINING */
 
 	/* Start retention register tracking for training firmware related registers */
 	ddrphy_phyinit_reginterface(starttrack, 0, 0);
 
 	/* (I) Load PHY Init Engine Image */
-	ddrphy_phyinit_i_loadpieimage(skip_training);
+	ddrphy_phyinit_i_loadpieimage(runtimeconfig.skip_train);
 
 	/*
 	 * Customize any CSR write desired to override values programmed by firmware or
@@ -103,9 +99,11 @@ int ddrphy_phyinit_sequence(bool skip_training)
 	 */
 	ddrphy_phyinit_usercustom_customposttrain();
 
-	if (runtimeconfig.reten) {
+	if (reten) {
 		/* Save value of tracked registers for retention restore sequence. */
-		/* ddrphy_phyinit_usercustom_saveretregs(); */
+		ddrphy_phyinit_usercustom_saveretregs();
+
+		runtimeconfig.reten = reten;
 	}
 
 	/* (J) Initialize the PHY to Mission Mode through DFI Initialization */
diff --git a/drivers/st/ddr/stm32mp2_ddr.c b/drivers/st/ddr/stm32mp2_ddr.c
index 341d2b7f1..da4dc348a 100644
--- a/drivers/st/ddr/stm32mp2_ddr.c
+++ b/drivers/st/ddr/stm32mp2_ddr.c
@@ -17,8 +17,6 @@
 
 #include <platform_def.h>
 
-#define SKIP_TRAINING_PARAM	false
-
 #define DDRDBG_FRAC_PLL_LOCK	U(0x10)
 
 #define DDRCTL_REG(x, y)					\
@@ -390,7 +388,18 @@ void stm32mp2_ddr_init(struct stm32mp_ddr_priv *priv,
 
 	disable_refresh(priv->ctl);
 
-	ret = ddrphy_phyinit_sequence(SKIP_TRAINING_PARAM);
+	if (false /* TODO stm32mp1_is_wakeup_from_standby() */) {
+		/* Initialize DDR by skipping training and disabling result saving */
+		ret = ddrphy_phyinit_sequence(true, false);
+
+		if (ret == 0) {
+			ret = ddrphy_phyinit_restore_sequence();
+		}
+	} else {
+		/* Initialize DDR including training and result saving */
+		ret = ddrphy_phyinit_sequence(false, true);
+	}
+
 	if (ret != 0) {
 		ERROR("DDR PHY init: Error %d\n", ret);
 		panic();
diff --git a/plat/st/stm32mp2/platform.mk b/plat/st/stm32mp2/platform.mk
index f4c3f3795..09ef45765 100644
--- a/plat/st/stm32mp2/platform.mk
+++ b/plat/st/stm32mp2/platform.mk
@@ -181,11 +181,14 @@ BL2_SOURCES			+=	drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_c_initphyconfig.c
 					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_isdbytedisabled.c				\
 					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_loadpieprodcode.c				\
 					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_mapdrvstren.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_progcsrskiptrain.c			\
 					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_reginterface.c				\
+					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_restore_sequence.c			\
 					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_sequence.c				\
 					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_softsetmb.c				\
 					drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_customposttrain.c	\
-					drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_custompretrain.c
+					drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_custompretrain.c	\
+					drivers/st/ddr/phy/phyinit/usercustom/ddrphy_phyinit_usercustom_saveretregs.c
 
 BL2_SOURCES			+=	drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_d_loadimem.c				\
 					drivers/st/ddr/phy/phyinit/src/ddrphy_phyinit_f_loaddmem.c				\
-- 
2.47.2

